### PR TITLE
Feature/add cluster force one shard option

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Add a startup option for starting cluster in (exclusive) one-shard mode
Also activates the cluster-one-shard rule permanently for enterprise builds

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/348

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *scripts/unittest shell_server_aql --cluster true*.

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -18,10 +18,10 @@ devel
   `distributeShardsLike` in case the prototype is a system collection, and the prototype
   should be moved to another server.
 
-* RCLone URL normalization
+* RClone URL normalization.
```
Comment (Simran-B) [inline CHANGELOG]: The correct capitalization is Rclone according to the official website
```suggestion
* Rclone URL normalization.
```

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6835/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6853/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6866/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/3.devel/job/arangodb-devel-matrix-nightly/230/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6877/

**Comment by jsteemann** [comment]
Tests blue

File: arangod/Graph/GraphManager.cpp [hunk 2]
```diff
@@ -97,18 +97,42 @@ OperationResult GraphManager::createCollection(std::string const& name, TRI_col_
...
+
+    uint64_t numberOfShards = 
+      arangodb::basics::VelocyPackHelper::readNumericValue<uint64_t>(options, StaticStrings::NumberOfShards, 0);
+ 
+    if (vocbase.server().getFeature<ClusterFeature>().forceOneShard()) {
+      // force a single shard with shards distributed like "_graph"
+      helper.add(StaticStrings::NumberOfShards, VPackValue(1));
+      char const* prototype = vocbase.shardingPrototype() == ShardingPrototype::Users ? TRI_COL_NAME_USERS : StaticStrings::GraphCollection.data();
+      helper.add(StaticStrings::DistributeShardsLike, VPackValue(prototype));
+    } else if (vocbase.name() == "single" && numberOfShards <= 1) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Graph/GraphManager.cpp]: Good catch :eyes:

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 3]
```diff
@@ -1090,23 +1090,45 @@ Result RestReplicationHandler::processRestoreCollectionCoordinator(
...
+    } else {
+      numberOfShards = numberOfShardsSlice.getUInt();
+    }
+    
+    if (_vocbase.sharding() == "single" && 
+        parameters.get(StaticStrings::DistributeShardsLike).isNone() &&
+        !_vocbase.IsSystemName(name) &&
+        numberOfShards <= 1) {
+      // shard like _graphs
+      char const* prototype = _vocbase.shardingPrototype() == ShardingPrototype::Users ? TRI_COL_NAME_USERS : StaticStrings::GraphCollection.data();
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/RestHandler/RestReplicationHandler.cpp]: Yes, makes sense. Will adjust the PR.

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6902/

File: arangod/Graph/GraphManager.cpp [hunk 4]
```diff
@@ -97,18 +97,42 @@ OperationResult GraphManager::createCollection(std::string const& name, TRI_col_
...
+    if (vocbase.server().getFeature<ClusterFeature>().forceOneShard()) {
+      // force a single shard with shards distributed like "_graph"
+      helper.add(StaticStrings::NumberOfShards, VPackValue(1));
+      helper.add(StaticStrings::DistributeShardsLike, VPackValue(vocbase.shardingPrototypeName()));
+    } else if (vocbase.sharding() == "single" && 
+               options.get(StaticStrings::DistributeShardsLike).isNone() &&
+               numberOfShards <= 1) {
+      helper.add(StaticStrings::NumberOfShards, VPackValue(1));
+      helper.add(StaticStrings::DistributeShardsLike, VPackValue(vocbase.shardingPrototypeName()));
+    }
```
Comment (danielhlarkin) [inline arangod/Graph/GraphManager.cpp]: Both branches of this if seem to be the same. Should we just combine the clauses?

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6911/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6919/

**Pull request merged**